src/Compilers/CSharp/Portable/Binder/SwitchBinder_Patterns.cs

```diff
@@ -218,6 +218,8 @@ internal override void BindPatternSwitchLabelForInference(CasePatternSwitchLabel
                         var caseLabelSyntax = (CaseSwitchLabelSyntax)node;
                         BoundConstantPattern pattern = sectionBinder.BindConstantPattern(
                             node, SwitchGoverningType, caseLabelSyntax.Value, node.HasErrors, diagnostics, out bool wasExpression);
+                        if (!pattern.HasErrors && caseLabelSyntax.Value is IdentifierNameSyntax name && name.Identifier.ContextualKind() == SyntaxKind.UnderscoreToken)
+                            diagnostics.Add(ErrorCode.WRN_CaseConstantNamedUnderscore, caseLabelSyntax.Value.Location);
```

Comment:
Consider extracting a local function for use here and below:
```
CheckUnderscore(caseLabelSyntax.Value, pattern.HasErrors, diagnostics);
```

Reply:
Why do we want to unconditionally provide a warning? Wouldn't it make more sense if there's no `_` in scope?

Reply:
This was an LDM decision. I think the reasoning is that everywhere else a pattern is accepted, `_` means a discard, unconditionally.

---
In reply to: [230201415](https://github.com/dotnet/roslyn/pull/30847#discussion_r230201415) [](ancestors = 230201415)